internal/staging/stdscript/scriptv0.go

```diff
@@ -0,0 +1,835 @@
+	pk, sigType := ExtractPubKeyHashAltDetailsV0(script)
+	return pk != nil && sigType == dcrec.STSchnorrSecp256k1
+}
+
+// ExtractScriptHashV0 extracts the script hash from the passed script if it is
+// a standard version 0 pay-to-script-hash script.  It will return nil
+// otherwise.
+func ExtractScriptHashV0(script []byte) []byte {
+	// Defer to consensus code.
+	return txscript.ExtractScriptHash(script)
+}
+
+// IsScriptHashScriptV0 returns whether or not the passed script is a standard
+// version 0 pay-to-script-hash script.
+func IsScriptHashScriptV0(script []byte) bool {
+	return ExtractScriptHashV0(script) != nil
+}
+
+// MultiSigDetailsV0 houses details extracted from a standard version 0 ECDSA
+// multisig script.
+type MultiSigDetailsV0 struct {
+	RequiredSigs int
+	NumPubKeys   int
+	PubKeys      [][]byte
+	Valid        bool
+}
+
+// ExtractMultiSigScriptDetailsV0 attempts to extract details from the passed
+// version 0 script if it is a standard ECDSA multisig script.  The returned
+// details struct will have the valid flag set to false otherwise.
+//
+// The extract pubkeys flag indicates whether or not the pubkeys themselves
+// should also be extracted and is provided because extracting them results in
+// an allocation that the caller might wish to avoid.  The PubKeys member of the
+// returned details struct will be nil when the flag is false.
+func ExtractMultiSigScriptDetailsV0(script []byte, extractPubKeys bool) MultiSigDetailsV0 {
+	// A multi-signature script is of the form:
+	//  NUM_SIGS PUBKEY PUBKEY PUBKEY ... NUM_PUBKEYS OP_CHECKMULTISIG
+
+	// The script can't possibly be a multisig script if it doesn't end with
+	// OP_CHECKMULTISIG or have at least two small integer pushes preceding it.
+	// Fail fast to avoid more work below.
+	if len(script) < 3 || script[len(script)-1] != txscript.OP_CHECKMULTISIG {
+		return MultiSigDetailsV0{}
+	}
+
+	// The first opcode must be a small integer specifying the number of
+	// signatures required.
+	const scriptVersion = 0
+	tokenizer := txscript.MakeScriptTokenizer(scriptVersion, script)
+	if !tokenizer.Next() || !txscript.IsSmallInt(tokenizer.Opcode()) {
+		return MultiSigDetailsV0{}
+	}
+	requiredSigs := txscript.AsSmallInt(tokenizer.Opcode())
+
+	// There must be at least one required signature.
+	if requiredSigs == 0 {
+		return MultiSigDetailsV0{}
+	}
+
+	// The next series of opcodes must either push public keys or be a small
+	// integer specifying the number of public keys.  It should be noted that
+	// although the consensus rules allow a higher maximum number of pubkeys,
+	// this intentionally further restricts the maximum number to what can be
+	// represented by a small integer push (up to a max of 16).
+	var numPubKeys int
+	var pubKeys [][]byte
+	if extractPubKeys {
+		pubKeys = make([][]byte, 0, txscript.MaxPubKeysPerMultiSig)
+	}
+	for tokenizer.Next() {
+		data := tokenizer.Data()
+		if !txscript.IsStrictCompressedPubKeyEncoding(data) {
+			break
```

Comment:
I understand what you're saying, and it is correct in this particular scenario that `data == nil` will signal the end of the pushed public keys because a small integer is special in that it doesn't behave as a normal data push.

However, I'm a bit hesitant to make that change here because the `CHECKMULTISIG` opcode actually supports up to 20 public keys, which would have to be encoded as `OP_DATA_1 20` and thus `data` would _not_ be `nil` in that case.  As the comment above this code says, we are intentionally further restricting it to 16 here for the purposes of standardness, so it does just so happen to work out such that the `data` will be `nil`.

I should probably also note that this code originally allowed it to go up to 20, which is why it is the way that it is.  The further restriction to 16 wasn't added until later because it results in smaller scripts and, in practice, we've never seen multisignature in the wild with anywhere near that number of public key anyway.  Also, ultimately we really want to move away from these inefficient and naive ECDSA signature in favor of threshold signatures anyway, but that's another matter.

I believe the existing `almost v0 multisig 1-of-2 -- mixed (un)compressed pubkeys` negative test covers the case you're worried about already, right?
```
		name:     "almost v0 multisig 1-of-2 -- mixed (un)compressed pubkeys",
		script:   p("1 DATA_65 0x%s DATA_33 0x%s 2 CHECKMULTISIG", pkUE, pkCO),
		wantType: STNonStandard,
```

Reply:
I see why the logic is as it is now, thanks.  With the `IsSmallInt` check that restricts to 16 pubkeys, handling a data push after the loop didn't make much sense, but it certainly would for a number encoded with a `OP_DATA_1 N`.  Thanks for explaining.

That test is totally the same path, yes.